trading_pair_azero/lib.rs

```diff
@@ -55,98 +60,195 @@ pub mod trading_pair_azero {
     }
 
     #[ink(event)]
     pub struct LiquidityPoolWithdrawal {
-        caller: AccountId,  // Address of the caller who initiated the liquidity withdrawal
-        shares_given: Balance,  // Amount of LP tokens (shares) being withdrawn
-        a0_given_amount: Balance,  // Amount of AZERO tokens given to the caller as part of the withdrawal
-        psp22_given_amount: Balance,  // Amount of PSP22 tokens given to the caller as part of the withdrawal
-        new_shares_balance: Balance  // Updated balance of LP tokens (shares) after the withdrawal
+        caller: AccountId, // Address of the caller who initiated the liquidity withdrawal
+        shares_given: Balance, // Amount of LP tokens (shares) being withdrawn
+        a0_given_amount: Balance, /* Amount of AZERO tokens given to the caller as part of the withdrawal */
+        psp22_given_amount: Balance, /* Amount of PSP22 tokens given to the caller as part of the withdrawal */
+        new_shares_balance: Balance, // Updated balance of LP tokens (shares) after the withdrawal
     }
 
     #[ink(event)]
     pub struct A0Swap {
-        caller: AccountId,  // Address of the caller who initiated the A0 token swap
-        a0_deposited_amount: Balance,  // Amount of AZERO tokens deposited by the caller for the swap
-        psp22_given_amount: Balance,  // Amount of PSP22 tokens given to the caller as part of the swap
-        psp22_given_to_vault: Balance,  // Amount of PSP22 tokens sent to the vault as part of the swap
+        caller: AccountId, // Address of the caller who initiated the A0 token swap
+        a0_deposited_amount: Balance, // Amount of AZERO tokens deposited by the caller for the swap
+        psp22_given_amount: Balance, /* Amount of PSP22 tokens given to the caller as part of the swap */
+        psp22_given_to_vault: Balance, /* Amount of PSP22 tokens sent to the vault as part of the swap */
     }
 
     #[ink(event)]
     pub struct PSP22Swap {
-        caller: AccountId,  // Address of the caller who initiated the PSP22 token swap
-        psp22_deposited_amount: Balance,  // Amount of PSP22 tokens deposited by the caller for the swap
-        a0_given_amount: Balance,  // Amount of AZERO tokens given to the caller as part of the swap
-        a0_given_to_vault: Balance,  // Amount of AZERO tokens sent to the vault as part of the swap
+        caller: AccountId, // Address of the caller who initiated the PSP22 token swap
+        psp22_deposited_amount: Balance, /* Amount of PSP22 tokens deposited by the caller for the swap */
+        a0_given_amount: Balance, // Amount of AZERO tokens given to the caller as part of the swap
+        a0_given_to_vault: Balance, // Amount of AZERO tokens sent to the vault as part of the swap
     }
 
+    impl PSP22 for TradingPairAzero {
+        #[ink(message)]
+        fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+            self.lp_tokens_allowances
```

Comment:
this field is also in psp22::Data, so if you are using the OB implementation, there is no need for this field, same for the balances field. Also, the implementation seems to be the same the one from OB, so another question is if you need this custom implementation (it is okay if it is kept, just a question to ask)

Reply:
We wish to use both allowances and both balances, yes :)